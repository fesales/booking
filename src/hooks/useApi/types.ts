export enum FetchStatus {
  ERROR = 'ERROR',
  PENDING = 'PENDING',
  SUCCESS = 'SUCCESS',
}

export interface FetchError {
  message: string;
}

interface PendingState<ApiResponse> {
  response: ApiResponse | FetchError | null;
  status: FetchStatus.PENDING;
}

export interface SuccessState<ApiResponse> {
  response: ApiResponse | null;
  status: FetchStatus.SUCCESS;
}

interface ErrorState {
  response: FetchError;
  status: FetchStatus.ERROR;
}

export type State<ApiResponse> = PendingState<ApiResponse> | SuccessState<ApiResponse> | ErrorState;

export enum Actions {
  CLEAR_STATE = 'CLEAR_STATE',
  REQUEST_LOADING = 'REQUEST_LOADING',
  REQUEST_SUCCESS = 'REQUEST_SUCCESS',
  REQUEST_ERROR = 'REQUEST_ERROR',
}

export interface ClearStateAction {
  type: Actions.CLEAR_STATE;
}

export interface RequestLoadingAction {
  type: Actions.REQUEST_LOADING;
}

export interface RequestSuccessAction<ApiResponse> {
  payload: ApiResponse;
  type: Actions.REQUEST_SUCCESS;
}

export interface RequestErrorAction {
  payload: FetchError;
  type: Actions.REQUEST_ERROR;
}

export type Action<ApiResponse> =
  | ClearStateAction
  | RequestErrorAction
  | RequestSuccessAction<ApiResponse>
  | RequestLoadingAction;
