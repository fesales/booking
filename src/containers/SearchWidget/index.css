.search-widget {
  background-color: #f3ce55;
  padding: 24px;
  text-align: left;
  max-width: 500px;
}

.search-widget-title {
  margin-top: 0;
}
