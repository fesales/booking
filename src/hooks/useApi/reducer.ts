import {
  Action,
  Actions,
  State,
  FetchStatus,
  RequestSuccessAction,
  RequestErrorAction,
  FetchError,
  RequestLoadingAction,
} from './types';

export const initialState = {
  response: null,
  status: FetchStatus.SUCCESS,
};

export const requestLoading = (): RequestLoadingAction => ({
  type: Actions.REQUEST_LOADING,
});

export const requestError = (error: FetchError): RequestErrorAction => ({
  payload: error,
  type: Actions.REQUEST_ERROR,
});

export const requestSuccess = <ApiResponse>(
  data: ApiResponse
): RequestSuccessAction<ApiResponse> => ({
  payload: data,
  type: Actions.REQUEST_SUCCESS,
});

const reducer = <ApiResponse>(
  state: State<ApiResponse> = initialState,
  action: Action<ApiResponse>
): State<ApiResponse> => {
  switch (action.type) {
    case Actions.REQUEST_ERROR: {
      return {
        ...state,
        response: action.payload,
        status: FetchStatus.ERROR,
      };
    }
    case Actions.REQUEST_LOADING: {
      return {
        ...state,
        status: FetchStatus.PENDING,
      };
    }
    case Actions.REQUEST_SUCCESS: {
      return {
        ...state,
        response: action.payload,
        status: FetchStatus.SUCCESS,
      };
    }
    default:
      return state;
  }
};

export default reducer;
