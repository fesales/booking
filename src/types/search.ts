export enum PlaceType {
  AIRPORT = 'A',
  CITY = 'C',
  STATION = 'T',
}

export interface PickUpLocation {
  city: string;
  country: string;
  name: string;
  region: string;
  placeType: PlaceType;
}

export interface SearchApiResponse {
  results: {
    isGooglePowered: boolean;
    docs: PickUpLocation[];
    numFound: number;
  };
}
