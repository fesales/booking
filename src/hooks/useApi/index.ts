import { useReducer, Reducer, useCallback } from 'react';
import reducer, {
  clearState,
  initialState,
  requestSuccess,
  requestLoading,
  requestError,
} from './reducer';
import { Action, State } from './types';
import { getUrlWithQueryParams, RequestQueryParams } from '../../utils/url';

export type UseApiReturnType<ApiResponse> = [
  State<ApiResponse>,
  {
    clearRequest: () => void;
    makeRequest: (requestQueryParams?: RequestQueryParams) => void;
  }
];

const useApi = <ApiResponse>(endpoint: string): UseApiReturnType<ApiResponse> => {
  const [state, dispatch] = useReducer<Reducer<State<ApiResponse>, Action<ApiResponse>>>(
    reducer,
    initialState
  );

  const makeRequest = useCallback(
    async (requestQueryParams?: RequestQueryParams) => {
      dispatch(requestLoading());

      const requestEndpoint = requestQueryParams
        ? getUrlWithQueryParams(endpoint, requestQueryParams)
        : endpoint;

      try {
        const response = await fetch(requestEndpoint);
        dispatch(requestSuccess(await response.json()));
      } catch (error) {
        dispatch(requestError(error));
      }
    },
    [endpoint]
  );

  const clearRequest = useCallback(() => {
    dispatch(clearState());
  }, []);

  return [state, { clearRequest, makeRequest }];
};

export default useApi;
