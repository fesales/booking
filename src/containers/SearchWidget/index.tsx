import React, { useMemo } from 'react';
import './index.css';
import debounce from 'lodash.debounce';
import SearchSuggestionsInput from '../../components/SearchSuggestionInput';
import useApi from '../../hooks/useApi';
import config from '../../config/dev';
import { SearchApiResponse, PickUpLocation } from '../../types/search';
import { FetchStatus } from '../../hooks/useApi/types';

export const SEARCH_INPUT_PLACEHOLDER = 'city, airport, station, region and district...';
export const MAX_RESULTS = 6;

const SearchWidget = () => {
  const [{ response, status }, makeRequest] = useApi<SearchApiResponse>(config.endpoints.search);

  const onSearchInputChanged = (searchTerm: string) => {
    if (searchTerm.length > 1) {
      makeRequest({
        number_of_results_required: MAX_RESULTS,
        search_term: searchTerm,
      });
    }
  };

  const debouncedSearchInputChanged = useMemo(() => debounce(onSearchInputChanged, 500), []);

  const searchResults: PickUpLocation[] | null = useMemo(() => {
    return status === FetchStatus.SUCCESS && response
      ? (response as SearchApiResponse).results.docs
      : null;
  }, [response, status]);

  return (
    <section className="search-widget">
      <h3 className="search-widget-title">Where are you going?</h3>
      <SearchSuggestionsInput
        label="Pick up location"
        onSearchTermChanged={debouncedSearchInputChanged}
        placeholder={SEARCH_INPUT_PLACEHOLDER}
        searchResults={searchResults}
      />
    </section>
  );
};

export default SearchWidget;
