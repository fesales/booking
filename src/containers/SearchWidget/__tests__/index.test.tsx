import React from 'react';
import { shallow } from 'enzyme';
import SearchWidget, { SEARCH_INPUT_PLACEHOLDER, MAX_RESULTS } from '..';
import useApi from '../../../hooks/useApi';
import { FetchStatus } from '../../../hooks/useApi/types';

jest.mock('../../../hooks/useApi', () => jest.fn(() => [{}]));
jest.mock('lodash.debounce', () => (func: () => void) => func);

describe('Search widget', () => {
  it('should render component', () => {
    const wrapper = shallow(<SearchWidget />);

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('SearchSuggestionsInput')).toHaveLength(1);
    expect(wrapper.find('SearchSuggestionsInput').prop('placeholder')).toEqual(
      SEARCH_INPUT_PLACEHOLDER
    );
  });

  describe('search results', () => {
    const mockedState = {
      response: {
        results: {
          docs: [{ city: '' }],
        },
      },
      status: FetchStatus.SUCCESS,
    };

    it('should pass search results when it is not a success status', () => {
      (useApi as jest.Mock).mockReturnValueOnce([mockedState]);

      const wrapper = shallow(<SearchWidget />);

      expect(wrapper.find('SearchSuggestionsInput').prop('searchResults')).toEqual(
        mockedState.response.results.docs
      );
    });

    it('should pass null to search results when it is not a success status', () => {
      const state = {
        ...mockedState,
        status: FetchStatus.ERROR,
      };

      (useApi as jest.Mock).mockReturnValueOnce([state]);

      const wrapper = shallow(<SearchWidget />);

      expect(wrapper.find('SearchSuggestionsInput').prop('searchResults')).toBeNull();
    });
  });

  describe('search input changed', () => {
    it('should make new request when search term length is greater than 2', () => {
      const makeRequest = jest.fn();
      const searchTerm = 'test';

      (useApi as jest.Mock).mockReturnValueOnce([{}, makeRequest]);

      const wrapper = shallow(<SearchWidget />);

      wrapper.find('SearchSuggestionsInput').simulate('searchTermChanged', searchTerm);

      expect(makeRequest).toHaveBeenCalledTimes(1);
      expect(makeRequest).toHaveBeenCalledWith({
        number_of_results_required: MAX_RESULTS,
        search_term: searchTerm,
      });
    });

    it('should not make new request when search term length is less than 2', () => {
      const makeRequest = jest.fn();
      const searchTerm = 't';

      (useApi as jest.Mock).mockReturnValueOnce([{}, makeRequest]);

      const wrapper = shallow(<SearchWidget />);

      wrapper.find('SearchSuggestionsInput').simulate('searchTermChanged', searchTerm);

      expect(makeRequest).toHaveBeenCalledTimes(0);
    });
  });
});
