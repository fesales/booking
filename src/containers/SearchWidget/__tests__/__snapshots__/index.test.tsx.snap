// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search widget should render component 1`] = `
<section
  className="search-widget"
>
  <h3
    className="search-widget-title"
  >
    Where are you going?
  </h3>
  <SearchSuggestionsInput
    isLoading={false}
    label="Pick up location"
    onSearchTermChanged={[Function]}
    placeholder="city, airport, station, region and district..."
    searchResults={null}
  />
</section>
`;
